### PR TITLE
Fix for EE price and stock asynchronous indexers

### LINKED ISSUE #297: addFinalPrice is using the price index which is not reliable
Steps to reproduce:
- Enable the queue
- Update a product, reindex
- go into the price index and delete references to that updated product
- add that product to the algolia queue and attempt to process the queue

Possible way to fix it:
- change the value of use_price_index thanks to catalog_product_collection_apply_limitations_before

### PR DESCRIPTION
**Pull request by JanPetr**
Fixes #297 

Magento EE offers a feature which asynchronously reindexes products' prices and stocks. Because of that the data indexed in Algolia doesn't have to correspond with data in Magento, because prices and stock might not be reindexed yet when the push to Algolia is performed.

This PR implements mechanism which checks if prices/stocks were already reindexed and therefore if the products can be pushed to Algolia.
If those weren't reindexed yet, those products will be skipped and reindexed in another run of queue runner indexer.

Those checks work only with indexing queue enabled.

### CODE REVIEW

File: app/code/community/Algolia/Algoliasearch/Helper/IndexChecker.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    private function getPendingProductIds($storeId)
+    {
+        if (isset($this->pendingProductIds) === false) {
+            $priceIndexes = $this->pendingPriceIndex($storeId);
+            $stockIndexes = $this->pendingStockIndex($storeId);
+
+            $pendingProductIds = array_unique(array_merge($priceIndexes, $stockIndexes));
+            $this->pendingProductIds = array_flip($pendingProductIds);
```
Comment (michaelwylde) [inline app/code/community/Algolia/Algoliasearch/Helper/IndexChecker.php]: Clever idea, much more efficient then my in_array method.

File: app/code/community/Algolia/Algoliasearch/Helper/IndexChecker.php [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        $pendingProductIds = $this->getPendingProductIds($storeId);
+
+        if (empty($pendingProductIds)) {
+            return;
+        }
+
+        foreach ($productIds as $id) {
+            if (isset($this->pendingProductIds[$id])) {
+                // Throw an exception - this exception is caught in Queue class and it'll retry next time Algolia is processed.
+                throw new Exception('Price index still pending for entity ID ' . $id);
```
Comment (michaelwylde) [inline app/code/community/Algolia/Algoliasearch/Helper/IndexChecker.php]: It is probably worth updating this exception message as you do not know if it is the stock or price index that is still pending. (If you wanted you could search for the price and stock index separately, however I joined them as the two arrays could have some overlaps.) A more generic message that I used in the logs was "Skipping, index still pending for entity id"

Comment (JanPetr) [inline app/code/community/Algolia/Algoliasearch/Helper/IndexChecker.php]: Great catch @michaelwylde, thanks!
In the end I created a custom exception and changed the message. Is it better?

**Comment by JanPetr** [comment]
Thanks a lot for a review @michaelwylde. Did you give it a try? :)

**Comment by michaelwylde** [comment]
@JanPetr  I have just tested it with varying combinations of the settings and they all worked as expected.

**Comment by fasaro** [comment]
You can reset the indexer through command line with indexer:reset command.

This will give the list of indexes name:

php bin/magento indexer:info

Output:

design_config_grid                       Design Config Grid
customer_grid                            Customer Grid
catalog_category_product                 Category Products
catalog_product_category                 Product Categories
catalog_product_price                    Product Price
catalog_product_attribute                Product EAV
catalogsearch_fulltext                   Catalog Search
cataloginventory_stock                   Stock
catalogrule_rule                         Catalog Rule Product
catalogrule_product                      Catalog Product Rule

This will give the list of indexes status:

php bin/magento indexer:status

Output:

Design Config Grid:                                Ready
Customer Grid:                                     Ready
Category Products:                                 Ready
Product Categories:                                Ready
Product Price:                                     Ready
Product EAV:                                       Ready
Catalog Search:                                    Ready
Stock:                                             Processing
Catalog Rule Product:                              Ready
Catalog Product Rule:                              Ready

If you want to reset all the indexes, you can run the following command:

php bin/magento indexer:reset

If you want to reset particular index (e.g. cataloginventory_stock), you can run the following command:

php bin/magento indexer:reset cataloginventory_stock

**Pull request merged**